File: backend/src/main/java/com/java/NBE4_5_1_7/domain/mail/EmailService.java
package com.java.NBE4_5_1_7.domain.mail;

import java.time.Instant;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.concurrent.ScheduledFuture;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.mail.javamail.JavaMailSender;
import org.springframework.mail.javamail.MimeMessageHelper;
import org.springframework.scheduling.TaskScheduler;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;
import org.thymeleaf.context.Context;
import org.thymeleaf.spring6.SpringTemplateEngine;

import jakarta.mail.MessagingException;
import jakarta.mail.internet.MimeMessage;
import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class EmailService {

	private final JavaMailSender mailSender;
	private final SpringTemplateEngine templateEngine;
	private final TaskScheduler taskScheduler;

	@Value("${mail.username}")
	private String fromEmail; // 보내는 이메일 주소

	@Value("${admin.email}")
	private String adminEmail; // 관리자의 이메일 주소

	@Async("asyncExecutor")
	public void sendChatNotification(String sender, String messageContent, String timestamp) {
		// ADMIN, SYSTEM 메시지 이메일 알림 제외
		if ("ADMIN".equals(sender) || "SYSTEM".equals(sender)) {
			return;
		}

		String formattedKSTTimestamp = convertToKST(timestamp);

		// 스레드 블로킹 없이 동작
		ScheduledFuture<?> scheduledFuture = taskScheduler.schedule(() -> {
			try {
				Thread.sleep(2000);

				Context context = new Context();
				context.setVariable("sender", sender);
				context.setVariable("content", messageContent);
				context.setVariable("timestamp", formattedKSTTimestamp);
				String htmlBody = templateEngine.process("email-template", context);

				MimeMessage mimeMessage = mailSender.createMimeMessage();
				MimeMessageHelper messageHelper = new MimeMessageHelper(mimeMessage, true);
				messageHelper.setFrom(fromEmail);
				messageHelper.setTo(adminEmail);
				messageHelper.setSubject("[DevPrep] 새로운 고객센터 문의가 도착했습니다.");
				messageHelper.setText(htmlBody, true);

				mailSender.send(mimeMessage);
			} catch (MessagingException e) {
				throw new RuntimeException(e);
			} catch (InterruptedException e) {
				Thread.currentThread().interrupt();
			}
		}, new java.util.Date(System.currentTimeMillis() + 2000));
	}

	private String convertToKST(String isoTimestamp) {
		try {
			Instant instant = Instant.parse(isoTimestamp);
			ZonedDateTime kstDateTime = instant.atZone(ZoneId.of("Asia/Seoul"));
			DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy년 MM월 dd일 a hh:mm")
					.withZone(ZoneId.of("Asia/Seoul"));
			return kstDateTime.format(formatter);
		} catch (Exception e) {
			return isoTimestamp; // 변환 실패 시 원본 반환
		}
	}
}

File: backend/src/main/java/com/java/NBE4_5_3_7/domain/mail/EmailService.kt
package com.java.NBE4_5_3_7.domain.mail

import jakarta.mail.MessagingException
import org.springframework.beans.factory.annotation.Value
import org.springframework.mail.javamail.JavaMailSender
import org.springframework.mail.javamail.MimeMessageHelper
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Service
import org.thymeleaf.context.Context
import org.thymeleaf.spring6.SpringTemplateEngine
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.concurrent.CompletableFuture
import java.util.concurrent.TimeUnit

@Service
class EmailService(
    private val mailSender: JavaMailSender,
    private val templateEngine: SpringTemplateEngine,
    @Value("\${mail.username}") private val fromEmail: String,
    @Value("\${admin.email}") private val adminEmail: String,
) {

    @Async("asyncExecutor")
    fun sendChatNotification(sender: String, messageContent: String?, timestamp: String) {
        if ("ADMIN" == sender || "SYSTEM" == sender) return

        val formattedKSTTimestamp = convertToKST(timestamp)

        CompletableFuture.runAsync({
            try {
                TimeUnit.SECONDS.sleep(2)

                val context = Context().apply {
                    setVariable("sender", sender)
                    setVariable("content", messageContent)
                    setVariable("timestamp", formattedKSTTimestamp)
                }

                val htmlBody = templateEngine.process("email-template", context)

                val mimeMessage = mailSender.createMimeMessage()
                val messageHelper = MimeMessageHelper(mimeMessage, true)
                messageHelper.setFrom(fromEmail)
                messageHelper.setTo(adminEmail)
                messageHelper.setSubject("[DevPrep] 새로운 고객센터 문의가 도착했습니다.")
                messageHelper.setText(htmlBody, true)

                mailSender.send(mimeMessage)
            } catch (e: MessagingException) {
                throw RuntimeException(e)
            } catch (e: InterruptedException) {
                Thread.currentThread().interrupt()
            }
        })
    }

    private fun convertToKST(isoTimestamp: String): String? {
        try {
            val instant = Instant.parse(isoTimestamp)
            val kstDateTime = instant.atZone(ZoneId.of("Asia/Seoul"))
            val formatter = DateTimeFormatter.ofPattern("yyyy년 MM월 dd일 a hh:mm")
                .withZone(ZoneId.of("Asia/Seoul"))
            return kstDateTime.format(formatter)
        } catch (e: Exception) {
            return isoTimestamp
        }
    }
}